kong/init.lua

```diff
@@ -147,6 +151,172 @@ local function load_plugins(kong_conf, dao)
   return sorted_plugins
 end
 
+
+local function mock_plugins(ctx, phase)
+  local mock    = ctx.mock     or {}
+  local plugins = mock.plugins or {}
+  local phase   = phase        or ngx.get_phase()
+
+  local BasePlugin = require "kong.plugins.base_plugin"
+
+  for plugin, plugin_conf in plugins_iterator(singletons.loaded_plugins, phase == "rewrite" or phase == "access") do
+    if plugin.handler[phase] ~= BasePlugin[phase] then
+      table.insert(plugins, {
+        name        = plugin.handler._name,
+        version     = plugin.handler.VERSION,
+        priority    = plugin.handler.PRIORITY,
+        no_consumer = plugin.schema.no_consumer,
+        phase       = phase,
+        config      = plugin_conf,
+      })
+    end
+  end
+
+  mock.plugins = plugins
+  ctx.mock     = mock
+end
+
+
+local function mock_consumer(ctx)
+  local consumer_id
+
+  local args = arguments.load({ multipart = false })
+  if args then
+    consumer_id = args.consumer
+    if type(consumer_id) == "table" then
+      consumer_id = consumer_id[1]
+    end
+  end
+
+  if consumer_id then
+    local consumer
+    local consumer_cache_key = singletons.dao.consumers:cache_key(consumer_id)
+
+    if type(consumer_id) == "string" or type(consumer_id) == "number" then
+      consumer_id = tostring(consumer_id)
+
+      if utils.is_valid_uuid(consumer_id) then
+        consumer = singletons.cache:get(
+          consumer_cache_key,
+          nil,
+          function(consumer_id)
+            return singletons.dao.consumers:find { id = consumer_id }
+          end,
+          consumer_id)
+
+      else
+        local result = singletons.dao.consumers:find_all { username = consumer_id }
+        if type(result) == "table" and type(result[1]) == "table" then
+          consumer =  result[1]
+        end
+
+        if not consumer then
+          result = singletons.dao.consumers:find_all { custom_id = consumer_id }
+          if type(result) == "table" and type(result[1]) == "table" then
+            consumer = result[1]
+          end
+        end
+      end
+    end
+
+    if consumer then
+      ctx.authenticated_consumer   = consumer
+      ctx.authenticated_credential = {
+        consumer_id = consumer.id
+      }
+
+      ngx.req.set_header(constants.HEADERS.CONSUMER_ID,        consumer.id)
+      ngx.req.set_header(constants.HEADERS.CONSUMER_CUSTOM_ID, consumer.custom_id)
+      ngx.req.set_header(constants.HEADERS.CONSUMER_USERNAME,  consumer.username)
```

Comment:
Looking forward to the day this is abstracted away from the plugins and we do not have to duplicate this logic here, nor in any of them...